ProxySQL not cleaning up file-handles for resolvable-but-not-reachable hosts

This is using Proxysql 2.0.15, and we've had two examples of this bug. We have been using a [mirroring configuration](https://proxysql.com/documentation/mirroring/) to test out different database server configurations against our production load, which has been quite useful. A couple of days ago we had to make the mirror server unreachable (but still resolvable) in order to diagnose some database-side bugs. Proxysql correctly started throwing errors about that:

```
2021-04-14 20:55:48 MySQL_Monitor.cpp:2404:monitor_ping(): [ERROR] Server example-database.us-east-3.rds.amazonaws.com:3306 missed 3 heartbeats, shunning it and killing all the connections. Disabling other checks until the node comes back online.
2021-04-14 20:55:50 MySQL_HostGroups_Manager.cpp:2446:get_random_MySrvC(): [ERROR] Hostgroup 5 has no servers available! Checking servers shunned for more than 1 second
```

That's fine, we expected that. What we didn't expect is what happened when it ran out of open file-handles.

```
$ sudo lsof -p 2035 |wc -l
102424
```

At that point the log started throwing different errors:

```
2021-04-15 16:02:44 sqlite3db.cpp:163:execute(): [ERROR] SQLITE error: unable to open database file --- INSERT INTO mysql_connections_hour SELECT timestamp/3600*3600 , MAX(Client_Connections_aborted), AVG(Client_Connections_connected), MAX(Client_Connections_created), MAX(Server_Connections_aborted), AVG(Server_Connections_connected), MAX(Server_Connections_created), MAX(ConnPool_get_conn_failure), MAX(ConnPool_get_conn_immediate), MAX(ConnPool_get_conn_success), MAX(Questions), MAX(Slow_queries), MAX(GTID_consistent_queries) FROM mysql_connections WHERE timestamp >= 1618498800 AND timestamp < 1618502400 GROUP BY timestamp/3600
2021-04-15 16:02:44 sqlite3db.cpp:163:execute(): [ERROR] SQLITE error: unable to open database file --- INSERT INTO myhgm_connections_hour SELECT timestamp/3600*3600 , MAX(MyHGM_myconnpoll_destroy), MAX(MyHGM_myconnpoll_get), MAX(MyHGM_myconnpoll_get_ok), MAX(MyHGM_myconnpoll_push), MAX(MyHGM_myconnpoll_reset) FROM myhgm_connections WHERE timestamp >= 1618498800 AND timestamp < 1618502400 GROUP BY timestamp/3600
2021-04-15 16:02:44 sqlite3db.cpp:163:execute(): [ERROR] SQLITE error: unable to open database file --- INSERT INTO mysql_query_cache_hour SELECT timestamp/3600*3600 , MAX(count_GET), MAX(count_GET_OK), MAX(count_SET), MAX(bytes_IN), MAX(bytes_OUT), MAX(Entries_Purged), AVG(Entries_In_Cache), AVG(Memory_bytes) FROM mysql_query_cache WHERE timestamp >= 1618498800 AND timestamp < 1618502400 GROUP BY timestamp/3600
2021-04-15 16:02:44 sqlite3db.cpp:163:execute(): [ERROR] SQLITE error: unable to open database file --- INSERT INTO system_cpu_hour SELECT timestamp/3600*3600 , SUM(tms_utime), SUM(tms_stime) FROM system_cpu WHERE timestamp >= 1618498800 AND timestamp < 1618502400 GROUP BY timestamp/3600
```

Without file-handles, it couldn't open the sqllite files and things went badly. When looking at the open files, they're all TCP sockets.
```
proxysql 2035 proxysql  469u     sock                0,8      0t0    282604 protocol: TCP
proxysql 2035 proxysql  470u     sock                0,8      0t0    282606 protocol: TCP
proxysql 2035 proxysql  471u     sock                0,8      0t0    294771 protocol: TCP
proxysql 2035 proxysql  472u     sock                0,8      0t0    467881 protocol: TCP
proxysql 2035 proxysql  473u     sock                0,8      0t0    467952 protocol: TCP
proxysql 2035 proxysql  474u     sock                0,8      0t0    370287 protocol: TCP
proxysql 2035 proxysql  475u     sock                0,8      0t0    460726 protocol: TCP
```

 Interestingly enough, this pathology doesn't happen if the database is **not** resolvable. We are working on getting an upgrade to 2.0.18 in place.

Another interesting side-effect: if you try to connect to the admin account while it is in this state, proxysql exceptions out and restarts.
```

pthread_create: Resource temporarily unavailable
2021-04-14 16:19:43 ProxySQL_Admin.cpp:4649:admin_main_loop(): [ERROR] Thread creation
proxysql: ProxySQL_Admin.cpp:4650: void* admin_main_loop(void*): Assertion `0' failed.
Error: signal 6:
proxysql(_Z13crash_handleri+0x2d)[0x4d743d]
/lib/x86_64-linux-gnu/libc.so.6(+0x354c0)[0x7f1e18e104c0]
/lib/x86_64-linux-gnu/libc.so.6(gsignal+0x38)[0x7f1e18e10438]
/lib/x86_64-linux-gnu/libc.so.6(abort+0x16a)[0x7f1e18e1203a]
/lib/x86_64-linux-gnu/libc.so.6(+0x2dbe7)[0x7f1e18e08be7]
/lib/x86_64-linux-gnu/libc.so.6(+0x2dc92)[0x7f1e18e08c92]
proxysql[0x588348]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x76ba)[0x7f1e19e596ba]
/lib/x86_64-linux-gnu/libc.so.6(clone+0x6d)[0x7f1e18ee24dd]
2021-04-14 16:19:44 main.cpp:1573:ProxySQL_daemonize_phase3(): [ERROR] ProxySQL crashed. Restarting!
```

--- a/lib/MySQL_Session.cpp
+++ b/lib/MySQL_Session.cpp
@@ -2628,6 +2628,14 @@ bool MySQL_Session::handler_again___status_CONNECTING_SERVER(int *_rc) {
 				previous_status.pop();
 			}
 			if (mybe->server_myds->myconn) {
+				// Created connection never reached 'connect_cont' phase, due to that
+				// internal structures of 'mysql->net' are not fully initialized.
+				// This induces a leak of the 'fd' associated with the socket
+				// opened by the library. To prevent this, we need to call
+				// `mysql_real_connect_cont` through `connect_cont`. This way
+				// we ensure a proper cleanup of all the resources when 'mysql_close'
+				// is later called. For more context see issue #3404.
+				mybe->server_myds->myconn->connect_cont(MYSQL_WAIT_TIMEOUT);
 				mybe->server_myds->destroy_MySQL_Connection_From_Pool(false);
 				if (mirror) {
 					PROXY_TRACE();

--- a/test/tap/tests/repro_3404-mysql_close_fd_leak.cpp
+++ b/test/tap/tests/repro_3404-mysql_close_fd_leak.cpp
@@ -1,0 +1,185 @@
+/**
+ * @file res_3404-mysql_close_fd_leak.cpp
+ * @brief This file is not a test to test to be executed,
+ *   it contains an isolated reproduction of issue #3404.
+ * @details This test contains and isolated reproduction of
+ *   issue #3404. For this, it uses the async 'libmariadbclient'
+ *   API in a fashion in which a file descriptor leak is created.
+ *
+ *   # Test usage
+ *
+ *   As the rest of tap tests this file requires serveral constants to
+ *   be supplied, for making use of this file it's required to execute
+ *   the following commands:
+ *
+ *   ```
+ *     $ source constants
+ *     $ sudo iptables -A OUTPUT -p tcp --dport $TAP_PORT -j DROP
+ *   ```
+ *
+ *   This way we make sure that the connection is going to try to
+ *   be performed and that the MySQL connection attempt is going
+ *   to fail with a timeout. Finally we launch the test file,
+ *   supplying the 'leak' parameter, to instruct the test to
+ *   use the flow that creates the leak:
+ *
+ *   ```
+ *     $ ./tests/repro_3404-mysql_close_fd_leak-t leak
+ *   ```
+ *
+ *   Check that the file descriptors are being leak:
+ *
+ *   ```
+ *     sudo lsof -p $(pgrep -f res_3404-socket_fd_not_closed) | wc -l
+ *   ```
+ *
+ *   We can also verify that nothing is being leak if we don't supply
+ *   the 'leak' parameter to the executable.
+ *
+ *   # Implementation details
+ *
+ *   For creating the leak this test creates the following flow:
+ *
+ *   1. Initialize the 'MYSQL' object using 'mysql_init'.
+ *   2. Attempt to start a connection with 'mysql_real_connect_start'.
+ *   3. Connection should timeout due to the server being unreachable.
+ *      Check usage.
+ *   4. We immediately call 'mysql_close'. Leaking a file descriptor.
+ *
+ */
+
+#include <cstring>
+#include <vector>
+#include <string>
+#include <stdio.h>
+#include <unistd.h>
+#include <poll.h>
+#include <sys/epoll.h>
+
+#include <mysql.h>
+#include <thread>
+
+#include "tap.h"
+#include "command_line.h"
+#include "utils.h"
+
+
+/* Helper function to do the waiting for events on the socket. */
+static int wait_for_mysql(MYSQL *mysql, int status) {
+	struct pollfd pfd;
+	int timeout, res;
+
+	pfd.fd = mysql_get_socket(mysql);
+	pfd.events =
+		(status & MYSQL_WAIT_READ ? POLLIN : 0) |
+		(status & MYSQL_WAIT_WRITE ? POLLOUT : 0) |
+		(status & MYSQL_WAIT_EXCEPT ? POLLPRI : 0);
+	if (status & MYSQL_WAIT_TIMEOUT)
+		timeout = 1000*mysql_get_timeout_value(mysql);
+	else
+		timeout = -1;
+	res = poll(&pfd, 1, timeout);
+	if (res == 0)
+		return MYSQL_WAIT_TIMEOUT;
+	else if (res < 0)
+		return MYSQL_WAIT_TIMEOUT;
+	else {
+		int status = 0;
+		if (pfd.revents & POLLIN) status |= MYSQL_WAIT_READ;
+		if (pfd.revents & POLLOUT) status |= MYSQL_WAIT_WRITE;
+		if (pfd.revents & POLLPRI) status |= MYSQL_WAIT_EXCEPT;
+		return status;
+	}
+}
+
+int main(int argc, char** argv) {
+	CommandLine cl;
+
+	if (cl.getEnv()) {
+		diag("Failed to get the required environmental variables.");
+		return -1;
+	}
+
+	bool create_leak = false;
+
+	if (argc == 2) {
+		std::string param { argv[1] };
+
+		if (param == "leak") {
+			create_leak = true;
+		}
+	}
+
+	for (int i = 0; i < 1000; i++) {
+		MYSQL* proxysql = mysql_init(NULL);
+		MYSQL* ret = NULL;
+		unsigned int timeout = 1;
+
+		mysql_options(proxysql, MYSQL_OPT_NONBLOCK, 0);
+		mysql_options(proxysql, MYSQL_OPT_CONNECT_TIMEOUT, (void *)&timeout);
+		mysql_ssl_set(proxysql, NULL, NULL, NULL, NULL, NULL);
+
+		int status = 0;
+
+		diag("Openning connection number: %d", i);
+
+		status =
+			mysql_real_connect_start(
+				&ret,
+				proxysql,
+				cl.host,
+				cl.username,
+				cl.password,
+				NULL,
+				cl.port,
+				NULL,
+				CLIENT_SSL
+			);
+
+		if (status == 0) {
+			fprintf(
+				stderr, "File %s, line %d, Error: %s\n",
+				__FILE__, __LINE__, mysql_error(proxysql)
+			);
+			return -1;
+		}
+
+		my_socket sockt = mysql_get_socket(proxysql);
+
+		int state = 0;
+		MYSQL* ret_mysql = NULL;
+
+		while (status) {
+			diag(":: Waiting for MySQL server on connection '%d'", i);
+			status = wait_for_mysql(proxysql, status);
+
+			// don't do finalize the connect! We directly call 'mysql_close'
+			// creating a leak in the already internally initalized 'fd'
+			// created by 'libmariadbclient'.
+			if (status == MYSQL_WAIT_TIMEOUT && create_leak) {
+				diag(":: Premature close in connection '%d', leaking 'fd'...", i);
+				break;
+			}
+
+			// even if we timeout, we call continue
+			diag(":: Calling 'mysql_real_connect_cont' with status: %d", status);
+			status = mysql_real_connect_cont(&ret_mysql, proxysql, status);
+		}
+
+		if (ret_mysql) {
+			diag(":: This is not expected, make sure the server is not reachable.");
+		} else {
+			// now we close
+			diag(
+				":: Calling 'mysql_close' after 'mysql_real_connect_cont' returned status: %d",
+				status
+			);
+			mysql_close(proxysql);
+		}
+
+		sleep(1);
+	}
+
+	return exit_status();
+}
+

